### PR TITLE
Add KSPSolveTranspose for PETSc

### PR DESCRIPTION
**Pull request by cmd8**

### CODE REVIEW

File: plugin/mpi/PETSc-code.hpp [hunk 1]
```diff
@@ -4998,7 +4998,9 @@ namespace PETSc {
     Global.Add("KSPSolve", "(", new PETSc::LinearSolver< Dmat >( ));
     Global.Add("KSPSolve", "(", new PETSc::LinearSolver< Dmat >(1));
     Global.Add("KSPSolve", "(", new PETSc::LinearSolver< Dmat >(1, 1));
-    if (!std::is_same< PetscScalar, PetscReal >::value)
+    if (std::is_same< PetscScalar, PetscReal >::value)
```
Comment (prj-) [inline plugin/mpi/PETSc-code.hpp:5001]: I don’t get the logic, what if you want to transpose solve (but not conjugate) in complex? The distinction must be made in  the `LinearSolver` function, not here, IMHO.

Comment (cmd8) [reply plugin/mpi/PETSc-code.hpp:5001]: I agree that the approach you suggest would be better. However, it is not possible to do a transpose solve at all (even in real) at the moment.

Comment (prj-) [reply plugin/mpi/PETSc-code.hpp:5001]: Yes, but if it is to be added, it should work in both arithmetics. It’s not that much extra work (but I’m AFK right now).

Comment (cmd8) [reply plugin/mpi/PETSc-code.hpp]: OK, I think my new commit addresses this.

File: plugin/mpi/PETSc-code.hpp [hunk 2]
```diff
@@ -3450,8 +3454,10 @@ namespace PETSc {
...
           PetscInt N, rbegin;
           PetscScalar* tmpIn, *tmpOut;
-          if (c != 3)
+          if (c != 3 && c != 5)
             KSPSolve(ptA->_ksp, x, y);
+          else if (c == 5)
+            KSPSolveTranspose(ptA->_ksp, x, y);
           else {
             ffassert(in != out);
             VecConjugate(x);
```
Comment (prj-) [inline plugin/mpi/PETSc-code.hpp:3463]:
```suggestion
          else {
            ffassert(in != out);
            if (c != 5) VecConjugate(x);
```

Comment (prj-) [reply plugin/mpi/PETSc-code.hpp:3463]: There is also another `if (c != 5)` that should appear below but I can't comment on that line.

File: plugin/mpi/PETSc-code.hpp [hunk 3]
```diff
@@ -3450,11 +3454,11 @@ namespace PETSc {
...
           PetscScalar* tmpIn, *tmpOut;
-          if (c != 3)
+          if (c != 3 && c != 5)
             KSPSolve(ptA->_ksp, x, y);
           else {
             ffassert(in != out);
-            VecConjugate(x);
+            if (c != 5) VecConjugate(x);
             KSPSolveTranspose(ptA->_ksp, x, y);
             VecConjugate(y);
```
Comment (prj-) [inline plugin/mpi/PETSc-code.hpp:3463]:
```suggestion
            if (c != 5) VecConjugate(y);
```

**Comment by cmd8** [comment]
Dear @prj-, I have noticed that `KSPSolveTranspose()` does not work properly on a MatNest (using `-ksp_type preonly -pc_type fieldsplit -pc_fieldsplit_type schur -pc_fieldsplit_schur_precondition self`. I get the correct solution for `KSPSolve()`, but not for `KSPSolveTranspose()`. Is this a bug or expected behavior?

**Comment by prj-** [comment]
Could you please share a MWE?

**Pull request merged**